pkg/store/proxy.go

```diff
@@ -263,346 +258,88 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
-					return
-				case rCh <- &recvResponse{r: r, err: err}:
-				}
-			}
-		}()
-
-		shardMatcher := shardInfo.Matcher(buffers)
-		defer shardMatcher.Close()
-		applySharding := shardInfo != nil && !storeSupportsSharding
-		if applySharding {
-			msg := "Applying series sharding in the proxy since there is not support in the underlying store"
-			level.Debug(logger).Log("msg", msg, "store", name)
+		if resp.GetWarning() != "" && (r.PartialResponseDisabled || r.PartialResponseStrategy == storepb.PartialResponseStrategy_ABORT) {
+			return status.Error(codes.Aborted, resp.GetWarning())
 		}
 
-		// The `defer` only executed when function return, we do `defer cancel` in for loop,
-		// so make the loop body as a function, release timers created by context as early.
-		handleRecvResponse := func() (next bool) {
-			frameTimeoutCtx, cancel := frameCtx(s.responseTimeout)
-			defer cancel()
-			var rr *recvResponse
-			select {
-			case <-ctx.Done():
-				s.handleErr(errors.Wrapf(ctx.Err(), "failed to receive any data from %s", s.name), done)
-				return false
-			case <-frameTimeoutCtx.Done():
-				s.handleErr(errors.Wrapf(frameTimeoutCtx.Err(), "failed to receive any data in %s from %s", s.responseTimeout.String(), s.name), done)
-				return false
-			case rr = <-rCh:
-			}
-
-			if rr.err == io.EOF {
-				close(done)
-				return false
-			}
-
-			if rr.err != nil {
-				s.handleErr(errors.Wrapf(rr.err, "receive series from %s", s.name), done)
-				return false
-			}
-			numResponses++
-			bytesProcessed += rr.r.Size()
-
-			if w := rr.r.GetWarning(); w != "" {
-				s.warnCh.send(storepb.NewWarnSeriesResponse(errors.New(w)))
-			}
-
-			if series := rr.r.GetSeries(); series != nil {
-				if applySharding && !shardMatcher.MatchesZLabels(series.Labels) {
-					return true
-				}
-
-				seriesStats.Count(series)
-
-				select {
-				case s.recvCh <- series:
-				case <-ctx.Done():
-					s.handleErr(errors.Wrapf(ctx.Err(), "failed to receive any data from %s", s.name), done)
-					return false
-				}
-			}
-			return true
+		if err := srv.Send(resp); err != nil {
+			return status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
 		}
-		for {
-			if !handleRecvResponse() {
-				return
-			}
-		}
-	}()
-	return s
-}
-
-func (s *streamSeriesSet) handleErr(err error, done chan struct{}) {
-	defer close(done)
-	s.closeSeries()
-
-	if s.partialResponse {
-		level.Warn(s.logger).Log("err", err, "msg", "returning partial response")
-		s.warnCh.send(storepb.NewWarnSeriesResponse(err))
-		return
 	}
-	s.errMtx.Lock()
-	s.err = err
-	s.errMtx.Unlock()
-}
-
-// Next blocks until new message is received or stream is closed or operation is timed out.
-func (s *streamSeriesSet) Next() (ok bool) {
-	s.currSeries, ok = <-s.recvCh
-	return ok
-}
 
-func (s *streamSeriesSet) At() (labels.Labels, []storepb.AggrChunk) {
-	if s.currSeries == nil {
-		return nil, nil
-	}
-	return s.currSeries.PromLabels(), s.currSeries.Chunks
+	return nil
```

Comment:
I think that's important one!

Reply:
Actually, it's not necessary because we convert *all* errors into warning responses and then the Series() function takes a decision whether to convert it into an error or not :upside_down_face: